File: src/com/suushiemaniac/cubing/alglib/move/Move.java
package com.suushiemaniac.cubing.alglib.move;

import com.suushiemaniac.cubing.alglib.move.modifier.Modifier;
import com.suushiemaniac.cubing.alglib.move.plane.Plane;
import com.suushiemaniac.cubing.alglib.transform.Transform;
import com.suushiemaniac.cubing.alglib.util.StringFormat;

public interface Move extends StringFormat {
    boolean merges(Move other);

    boolean cancels(Move other);

    boolean mayAppend(Move other);

    Plane getPlane();

    Modifier getModifier();

    Move merge(Move other);

    int getDepth();

    Move inverse();

    Move transform(Transform transform);

    boolean equals(Object other);

    int hashCode();
}

File: src/main/kotlin/com/suushiemaniac/cubing/alglib/move/Move.kt
package com.suushiemaniac.cubing.alglib.move

import com.suushiemaniac.cubing.alglib.move.modifier.Modifier
import com.suushiemaniac.cubing.alglib.move.plane.Plane
import com.suushiemaniac.cubing.alglib.transform.Transform
import com.suushiemaniac.cubing.alglib.util.StringFormat

interface Move : StringFormat {
    val plane: Plane
    val modifier: Modifier
    val depth: Int

    fun merges(other: Move): Boolean

    fun cancels(other: Move): Boolean

    fun mayAppend(other: Move): Boolean

    fun merge(other: Move): Move // FIXME nullability

    fun inverse(): Move // FIXME nullability (Megaminx…)

    fun transform(transform: Transform): Move = transform.transform(this)
}
